### PR TITLE
HTTP auth API

### LINKED ISSUE #84: Support username / password and CURLOPT_HTTPAUTH settings.
Seems like an oversight that username / password is not supported at all.

### PR DESCRIPTION
**Pull request by sagebind**
Add a new high-level API for specifying credentials and using one or more HTTP authentication schemes.

Fixes #84.

### CODE REVIEW

File: src/auth.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        *self = self.clone() | rhs;
+    }
+}
+
+impl SetOpt for Authentication {
+    fn set_opt<H>(&self, easy: &mut curl::easy::Easy2<H>) -> Result<(), curl::Error> {
+        #[cfg(feature = "spnego")]
+        {
+            if self.contains(Authentication::negotiate()) {
+                easy.username(":")?;
```
Comment (abatkin) [inline src/auth.rs]: This works on Linux, but does not work on Windows (presumably something different in the SSPI codepath?). Instead, set to the empty string (or whatever), but you must also explicity set `password()` to something (i.e. empty string) in order to allow negotiate to work properly on Windows.

File: src/auth.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        easy.http_auth(&self.as_auth())
+    }
+}
+
+impl SetOpt for Proxy<Authentication> {
+    fn set_opt<H>(&self, easy: &mut curl::easy::Easy2<H>) -> Result<(), curl::Error> {
+        #[cfg(feature = "spnego")]
+        {
+            if self.0.contains(Authentication::negotiate()) {
+                easy.proxy_username(":")?;
```
Comment (abatkin) [inline src/auth.rs]: See above

File: src/auth.rs (hunk 1 above)
Comment (sagebind) [reply src/auth.rs]: Interesting. In Windows CI it appears that the `Authorization` header is being populated with _something_ (see the Windows-specific test in `tests/auth.rs`). But other than that I have no way to test this feature to see if it works. :cry: Does the empty string break it on Linux (meaning we must have platform-specific codepaths)?

Also, is this documented somewhere? I couldn't find much on this.

Comment (abatkin) [reply src/auth.rs]: From looking at both the cURL code and the documentation, I can tell that *something* needs to be set in order for Negotiate to work, but in some places, it is referred to as the username and in others as the password. On the command line, you would normally pass `-u username:password` (or more specifically for negotiate, "`-u :`"), which I suspect is how you ended up with `":"`.

But to specifically answer your question: As far as I can tell, both Windows and Linux don't care what you set for the username/passwords when doing Negotiate. For some reason it works on Linux if you *only* set username, but Windows wants *both* username and password. So you shouldn't need any platform-specific code, just set both username and password to the empty string on all platforms (at least that's how I did my testing, and that worked).

(of course, the fun thing is that for Negotiate, both the username and password are always ignored, setting them just "turns on" cURL's ability to do *any* authentication)

Comment (sagebind) [reply src/auth.rs]: OK, that makes sense. I'll update the code to reflect that.

**Pull request merged**